narwhals/typing.py

```diff
@@ -84,18 +83,6 @@ def cast(self, dtype: DType) -> Self: ...
 
 class CompliantNamespace(Protocol, Generic[CompliantSeriesT_co]):
     def col(self, *column_names: str) -> CompliantExpr[CompliantSeriesT_co]: ...
-    def _create_expr_from_callable(
-        self,
-        # Should be
-        # Callable[[CompliantDataFrame], Sequence[CompliantSeriesT_co]]  # noqa: ERA001
-        # but we can't have that because typing rules require an invariant type here
-        func: Callable[[Any], Sequence[Any]],
```

Comment:
@MarcoGorelli is this the same issue I had here?

https://github.com/vega/altair/blob/9002472d65875a8486de920e4cc585420efdd276/tools/schemapi/schemapi.py#L851-L885

If so, maybe these are helpful?

- https://typing.readthedocs.io/en/latest/spec/protocol.html#generic-protocols
- https://typing.readthedocs.io/en/latest/spec/constructors.html#specialization-during-construction

Reply:
thanks! could be, but in the end not all Namespaces in Narwhals have `_create_expr_from_callable`, it's only in `_pandas_like` and `_arrow` (because they don't have an expression-like object in their API)

So for now it's just a pandas-like / pyarrow thing, and not required of all backends, so it's not in CompliantNamespace

Reply:
Got it, so a `CompliantNamespace` would need to support either:
- an expression-like protocol
- `_create_expr_from_callable`

But I'm assuming they are mutually exclusive?

Reply:
`CompliantNamespace` needs to support a `col` method, which returns a `CompliantExpr` (which in turn needs to implement whole bunch of other  methods)

If a library doesn't have a concept of expressions, then their `CompliantExpr` objects would be backed by callables which just reuse their `CompliantSeries` implementation. To do that, they can use the `reuse_series_implementation` function in `_expression_parsing.py`, which expects the `CompliantExpr` object to also implement `_create_expr_from_callable`

Reply:
Thanks for explaining @MarcoGorelli